contributors/design-proposals/node/troubleshoot-running-pods.md

```diff
@@ -235,37 +249,28 @@ The status of a Debug Container is reported in a new field in `v1.PodStatus`:
 ```
 type PodStatus struct {
         ...
-        DebugStatuses []DebugStatus
-}
-
-type DebugStatus struct {
-        Name string
-        Command []string
-        Args []string
-        // Set only for Debug Containers
-        DebugContainerStatus v1.ContainerStatus
+        EphemeralContainerStatuses []v1.ContainerStatus
 }
 ```
 
-Initially this will be populated only for Debug Containers, but there's interest
-in tracking status for traditional exec in a similar manner. Ideally we can
-report both types of user intervention into a container with a single new type.
+This status is only populated for Debug Containers, but there's interest in
+tracking status for traditional exec in a similar manner.
 
-Note that `Command` and `Args` must be tracked in the status object because
-there is no spec for Debug Containers or exec. These must either be made
+Note that `Command` and `Args` would have to be tracked in the status object
+because there is no spec for Debug Containers or exec. These must either be made
 available by the runtime or tracked by the kubelet. For Debug Containers this
```

Comment:
doesn't the whole ephemeral container have to be tracked by the kubelet?

Reply:
why just the command and args? admission uses the definitions of containers to determine whether a user can do things like exec/attach to those containers ("cannot exec into a pod with containers running as root", etc). Container status alone (`EphemeralContainerStatuses []v1.ContainerStatus`) wouldn't include enough information about the ephemeral containers to make that determination.

what if the submitted ephemeral container and its status were both made available in pod status?

```
type PodStatus struct {
  ...
  EphemeralContainers []v1.EphemeralContainer
}

type EphemeralContainer struct {
  Spec v1.Container
  Status v1.ContainerStatus
}
```

Reply:
The kubelet currently has no method of keeping state of the v1.Container across restarts. After a restart, it can reconstruct a minimum v1.Container based on runtime labels, though. This could (but currently doesn't) include Command and Args, which can be queried from the run time.

Including the spec in the PodStatus would be pretty great, though, so it might be worth solving. @lavalamp what do you think about this?

Reply:
Remind me why this got reopened for discussion? I feel like if we keep going down this route we will end up at my original position of adding a `type` of `ephemeral` to containers, and allowing users to add or remove such containers from the spec.

Reply:
I think sig-auth just wants to make this works well with admission controllers and has a strong preference for using v1.Container, [for many good reasons](https://github.com/kubernetes/community/pull/649/files#r142722579).

@liggitt This is an important feature to sig-node, but we've been stuck for about a year. If a v1.Container is available to admission controllers for the creation request, can you live with only ContainerStatus for attach, at least for alpha? We don't need to allow exec.

Reply:
> The kubelet currently has no method of keeping state of the v1.Container across restarts.

I expected the kubelet to:
1. update pod status with the ephemeral container definition/status as it created the container
2. update the ephemeral container status by name from observed container state during normal sync
3. remove the ephemeral container from the pod status when the ephemeral container was removed

None of that requires the kubelet remembering the ephemeral container definition beyond the initial create, right?

> If a v1.Container is available to admission controllers for the creation request, can you live with only ContainerStatus for attach, at least for alpha? We don't need to allow exec.

Two thoughts:
1. attach (which is the intended use case here) requires almost identical protections to exec ("cannot attach to a pod with containers running as root", etc). If the create of the ephemeral container and the attach to it are separate API calls, but the container definition is unavailable at attach time, I'm unsure how that would be implementable.
2. the structure of an ephemeral container field in the pod status could not be changed [in an incompatible way](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#adding-unstable-features-to-stable-versions) (*if the schema of the alpha field must change in an incompatible way, a new field name must be used*), so if we started with `EphemeralContainerStatuses []v1.ContainerStatus`, that couldn't grow to include the container definition as alpha progressed.